graphql_compiler/schema_transformation/split_query.py

```diff
@@ -34,6 +39,340 @@ def __init__(self, query_ast):
+        type_info.leave(selection)
+    type_info.leave(ast.selection_set)
+
+    if made_changes:
+        ast_copy = copy(ast)
+        ast_copy.selection_set = SelectionSet(selections=new_selections)
+        return ast_copy
+    else:
+        return ast
+
+
+def _get_child_query_node_and_out_name(ast, type_info, child_field_name,
+                                       intermediate_out_name_assigner):
+    """Create a query node out of ast, return node and unique out_name on field with input name.
+
+    Create a new document out of the input AST, that has the same structure as the input. For
+    instance, if the input AST can be represented by
+        out_Human {
+          name
+        }
+    where out_Human is a vertex field going to type Human, the resulting document will be
+        {
+          Human {
+            name
+          }
+        }
+    If the input AST starts with a type coercion, the resulting document will start with the
+    coerced type, rather than the original union or interface type.
+
+    The output child_node will be wrapped around this new Document. In addition, if no field
+    of child_field_name currently exists, such a field will be added. If there is no @output
+    directive on this field, a new @output directive will be added.
+
+    Args:
+        ast: type Node, representing the AST that we're using to build a child node. It is not
+             modified by this function
+        type_info: TypeInfo, at the location of the ast
+        child_field_name: str. If no field of this name currently exists as a part of the root
+                          selections of the input AST, a new field will be created in the AST
+                          contained in the output child query node
+        intermediate_out_name_assigner: IntermediateOutNameAssigner, which will be used to
+                                        generate an out_name, if it's necessary to create a
+                                        new @output directive
+
+    Returns:
+        Tuple[SubQueryNode, str], the child sub query node wrapping around the input ast, and
+        the out_name of the @output directive uniquely identifying the field used or stitching
+        in this sub query node
+    """
+    child_type_name, child_selections = _get_child_type_and_selections(ast, type_info)
+    # Get existing field with name in child
+    child_property_field = _get_property_field(child_selections, child_field_name, [])
+    # Add @output if needed, record out_name
+    child_output_name = _get_out_name_optionally_add_output(
+        child_property_field, intermediate_out_name_assigner
+    )
+    # Get new child_selections
+    child_selections = _replace_or_insert_property_field(child_selections, child_property_field)
+    # Wrap around
+    # NOTE: if child_type_name does not actually exist as a root field (not all types are
+    # required to have a corresponding root vertex field), then this query will be invalid
```

Comment:
Nice catch! Is there some validation we can perform when merging schemas to ensure this is impossible at splitting time?

Reply:
If all object and interface types have a corresponding root vertex field, then this is impossible. But I'm not sure if that's true or should be enforced.

Reply:
I think it's probably a good idea to enforce that if a given cross-schema edge is attempted to be defined, that its endpoint types and their subtypes exist in the corresponding schemas' root query types.